File: ruby/scale-generator/scale_generator.rb
# frozen_string_literal: true

class Scale
  FLAT_SCALE_TONICS = %w[F Bb Eb Ab Db Gb d g c f bb eb].freeze
  private_constant :FLAT_SCALE_TONICS
  FLAT_SCALE = %w[C Db D Eb E F Gb G Ab A Bb B].freeze
  private_constant :FLAT_SCALE
  INITIAL_SCALE_INDEX = 0
  private_constant :INITIAL_SCALE_INDEX
  INTERVALS = { "m" => 1, "M" => 2, "A" => 3 }.freeze
  private_constant :INTERVALS
  NAME_FORMAT = "%<tonic>s %<scale_name>s"
  private_constant :NAME_FORMAT
  SHARP_SCALE = %w[C C# D D# E F F# G G# A A# B].freeze
  private_constant :SHARP_SCALE

  def initialize(tonic, scale_name, intervals = nil)
    @tonic = tonic.capitalize
    @scale_name = scale_name
    @intervals = intervals
    @scale = FLAT_SCALE_TONICS.include?(tonic) ? FLAT_SCALE : SHARP_SCALE
  end

  def name
    format(NAME_FORMAT, tonic: tonic, scale_name: scale_name)
  end

  def pitches
    tonic_scale =
      scale
      .index(tonic)
      .then { |index| scale.rotate(index) }
    return tonic_scale unless intervals

    generate_pitches(tonic_scale)
  end

  private

  attr_reader :tonic, :scale_name, :intervals, :scale

  def generate_pitches(tonic_scale)
    intervals
      .chars
      .each
      .with_object(tonic_scale)
      .reduce([INITIAL_SCALE_INDEX, []], &method(:add_tonic_scale))
      .last
  end

  def add_tonic_scale((last_index, acc), (interval, tonic_scale))
    acc << tonic_scale[last_index]
    last_index += INTERVALS[interval]
    [last_index, acc]
  end
end

File: javascript/scale-generator/scale-generator.js
export class Scale {
  static get SHARP_SCALE() {
    return ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]
  }

  static get FLAT_SCALE() {
    return ["C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B"]
  }

  static get FLAT_SCALE_TONICS() {
    return ["F", "Bb", "Eb", "Ab", "Db", "Gb", "d", "g", "c", "f", "bb", "eb"]
  }

  static get SEMITONE_STEPS() {
    return { m: 1, M: 2, A: 3 }
  }

  constructor(tonic) {
    this.tonic = this.capitalize(tonic)
    this.scale =
      Scale.FLAT_SCALE_TONICS.includes(tonic)
        ? Scale.FLAT_SCALE
        : Scale.SHARP_SCALE
  }

  chromatic() {
    const scale = this.scale
    const index = scale.indexOf(this.tonic)
    return this.rotate(scale, index)
  }

  interval(intervals) {
    const chromaticScale = this.chromatic()

    return (
      intervals
        .split("")
        .reduce(this.addChromaticScale(chromaticScale), [0, []])[1]
    )
  }

  capitalize(string) {
    return string.charAt(0).toUpperCase() + string.slice(1)
  }

  rotate(list, count) {
    if (count < 1) {
      return list
    }

    const head = list.shift()
    list.push(head)
    return this.rotate(list, count - 1)
  }

  addChromaticScale(chromaticScale) {
    return ([lastIndex, acc], step) => {
      acc.push(chromaticScale[lastIndex])
      lastIndex += Scale.SEMITONE_STEPS[step]
      return [lastIndex, acc]
    }
  }
}
